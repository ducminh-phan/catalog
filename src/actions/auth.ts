import { createAction, createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";

export interface LoginRequest {
  username: string;
  password: string;
}

export interface LoginResponse {
  access_token: string;
}

export const login = createAsyncThunk("auth/login", (body: LoginRequest) => {
  return Request.post<LoginResponse>("login", body);
});

export interface RegisterRequest {
  username: string;
  password: string;
  email: string;
  name: string;
}

export interface RegisterResponse {
  access_token: string;
}

export const register = createAsyncThunk(
  "auth/register",
  (body: RegisterRequest) =>
    Request.post<RegisterResponse>("registrations", body),
);

export const logout = createAction("auth/logout");
