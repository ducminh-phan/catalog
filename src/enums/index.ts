export const ITEMS_PER_PAGE = 2;

export const CATEGORIES_PER_PAGE = 2;
