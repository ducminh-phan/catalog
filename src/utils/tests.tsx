import { StylesProvider } from "@material-ui/styles";
import { render as rtlRender, RenderResult } from "@testing-library/react";
import { Rule, StyleSheet } from "jss";
import React, { ComponentType, ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";

import Notification from "components/Notification";
import { AuthProvider } from "contexts/auth";
import { NotificationProvider } from "contexts/notification";
import store from "store";

import Request from "./api";
import { UserResponse } from "./auth";
import storage from "./storage";

// Custom logic to make Material UI class names consistent in snapshots
const generateClassName = (
  rule: Rule,
  styleSheet?: StyleSheet<string>,
): string => `${styleSheet?.options.classNamePrefix}-${rule.key}`;

const render = (ui: ReactElement, renderOptions = {}): RenderResult => {
  const Wrapper = ({ children }: { children: ReactNode }): ReactElement => {
    return (
      <StylesProvider generateClassName={generateClassName}>
        <Provider store={store}>
          <NotificationProvider>
            <AuthProvider>{children}</AuthProvider>
            <Notification />
          </NotificationProvider>
        </Provider>
      </StylesProvider>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper as ComponentType, ...renderOptions });
};

export const setupUser = (): void => {
  storage.setToken("accessToken");
  expect(storage.getToken()).toBe("accessToken");
  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<UserResponse> =>
      Promise.resolve({
        email: "minhπ",
        username: "minhπ",
        created: "",
        updated: "",
        name: "minhπ",
        id: 0,
      }),
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
