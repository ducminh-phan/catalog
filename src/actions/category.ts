import { createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";
import * as types from "utils/types";

export type FetchCategoriesResponse = {
  totalCategories: number;
  categories: types.Category[];
};

export const fetchCategories = createAsyncThunk(
  "categories/fetch",
  ({ offset, limit }: types.PaginationParams) =>
    Request.get<FetchCategoriesResponse>(
      `categories?offset=${offset}&limit=${limit}`,
    ),
);

export type AddCategoryRequest = {
  name: string;
  description: string;
};

export const addCategory = createAsyncThunk(
  "categories/add",
  (payload: AddCategoryRequest) =>
    Request.post<types.Category>("categories", payload),
);
