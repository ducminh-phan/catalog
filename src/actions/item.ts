import { createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";
import * as types from "utils/types";

type FetchItemsRequest = types.PaginationParams & {
  categoryId?: string;
};

export type FetchItemsResponse = {
  totalItems: number;
  items: types.Item[];
};

export const fetchItems = createAsyncThunk(
  "items/fetch",
  ({ categoryId, offset, limit }: FetchItemsRequest) =>
    Request.get<FetchItemsResponse>(
      `categories/${categoryId}/items?offset=${offset}&limit=${limit}`,
    ),
);

type EditItemsRequest = {
  categoryId?: string;
  itemId: string;
  data: {
    name: string;
    description: string;
    price: number;
  };
};

type EditItemsResponse = types.Item;

export const editItem = createAsyncThunk(
  "items/edit",
  ({ categoryId, itemId, data }: EditItemsRequest) =>
    Request.put<EditItemsResponse>(
      `categories/${categoryId}/items/${itemId}`,
      data,
    ),
);

type AddItemRequest = {
  categoryId?: string;
  data: {
    name: string;
    description: string;
    price: number;
  };
};

type AddItemResponse = types.Item;

export const addItem = createAsyncThunk(
  "item/add",
  ({ categoryId, data }: AddItemRequest) =>
    Request.post<AddItemResponse>(`categories/${categoryId}/items`, data),
);

type DeleteItemRequest = {
  categoryId?: string;
  itemId: string;
};

export const deleteItem = createAsyncThunk(
  "item/delete",
  ({ categoryId, itemId }: DeleteItemRequest) =>
    Request.delete(`categories/${categoryId}/items/${itemId}`),
);
