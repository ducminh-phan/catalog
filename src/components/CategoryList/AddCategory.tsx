import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from "@material-ui/core";
import AddIcon from "@material-ui/icons/Add";
import { unwrapResult } from "@reduxjs/toolkit";
import { makeRequired, makeValidate, TextField } from "mui-rff";
import React, { ReactNode } from "react";
import { Form } from "react-final-form";
import { connect, ConnectedProps } from "react-redux";
import * as yup from "yup";

import { addCategory } from "actions/category";
import { useNotification } from "contexts/notification";

const schema = yup.object().shape({
  name: yup.string().required().min(5),
  description: yup.string().required(),
});

const validate = makeValidate(schema);

const required = makeRequired(schema);

type FormData = yup.InferType<typeof schema>;

const mapDispatchToProps = {
  addCategory,
};

const connector = connect(null, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux;

const AddCategory = (props: Props): React.ReactElement => {
  // eslint-disable-next-line no-shadow
  const { addCategory } = props;
  const [open, setOpen] = React.useState(false);
  const { setNotification } = useNotification();

  const handleOpen = (): void => {
    setOpen(true);
  };

  const handleClose = (): void => {
    setOpen(false);
  };

  const handleAdd = async (data: FormData): Promise<void> => {
    const resultAction = await addCategory(data);

    handleClose();

    try {
      unwrapResult(resultAction);
      setNotification({
        severity: "success",
        message: "Add category successfully",
      });
    } catch (e) {
      setNotification({
        severity: "error",
        message: e.message ?? "Add category failed",
      });
    }
  };

  return (
    <>
      <Button color="inherit" onClick={handleOpen}>
        <AddIcon />
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="category-add-form-title"
        fullWidth
      >
        <DialogTitle id="category-add-form-title">Add Category</DialogTitle>
        <DialogContent>
          <Form
            onSubmit={handleAdd}
            validate={validate}
            render={({ handleSubmit, invalid }): ReactNode => (
              <form onSubmit={handleSubmit} noValidate>
                <TextField label="Name" name="name" required={required.name} />
                <TextField
                  label="Description"
                  name="description"
                  required={required.description}
                />
                <DialogActions>
                  <Button type="button" onClick={handleClose}>
                    Cancel
                  </Button>
                  <Button color="primary" type="submit" disabled={invalid}>
                    Submit
                  </Button>
                </DialogActions>
              </form>
            )}
          />
        </DialogContent>
      </Dialog>
    </>
  );
};

export default connector(AddCategory);
