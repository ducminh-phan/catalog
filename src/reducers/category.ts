import { createReducer } from "@reduxjs/toolkit";
import { normalize } from "normalizr";

import { fetchCategories } from "actions/category";
import * as schemas from "utils/schemas";
import * as types from "utils/types";

interface CategoryState {
  totalCategories: number;
  categories: {
    [key: string]: types.Category;
  };
  ids: string[];
}

const initialState: CategoryState = {
  totalCategories: 0,
  categories: {},
  ids: [],
};

export default createReducer(initialState, (builder) =>
  builder.addCase(fetchCategories.fulfilled, (state, action) => {
    const { payload } = action;
    const normalized = normalize(payload.categories, schemas.categories);

    return {
      totalCategories: payload.totalCategories,
      categories: normalized.entities.categories ?? state.categories,
      ids: normalized.result,
    };
  }),
);
