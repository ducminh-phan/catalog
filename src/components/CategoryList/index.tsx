import { Box, List, ListItem, ListItemText } from "@material-ui/core";
import Pagination from "@material-ui/lab/Pagination";
import React, { ChangeEvent, ReactElement, useEffect, useState } from "react";
import { connect, ConnectedProps } from "react-redux";

import { fetchCategories } from "actions/category";
import { RootState } from "reducers";
import * as types from "utils/types";

const mapDispatchToProps = {
  fetchCategories,
};

const connector = connect(
  ({ category }: RootState) => ({ category }),
  mapDispatchToProps,
);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux;

const limit = 2;

const CategoryList = (props: Props): ReactElement => {
  // eslint-disable-next-line no-shadow
  const { fetchCategories, category } = props;

  const [offset, setOffset] = useState(0);
  const [page, setPage] = useState(1);

  useEffect(() => {
    fetchCategories({ offset, limit });
  }, [fetchCategories, offset]);

  const { totalCategories, categories } = category;
  const pageCount = Math.ceil(totalCategories / limit);

  const handleChangePage = (
    event: ChangeEvent<unknown>,
    value: number,
  ): void => {
    setPage(value);
    setOffset(limit * (value - 1));
  };

  return (
    <div>
      <List>
        {Object.values(categories).map((c: types.Category) => (
          <ListItem key={c.id}>
            <ListItemText primary={c.name} secondary={c.description} />
          </ListItem>
        ))}
      </List>
      <Box textAlign="center">
        <Pagination count={pageCount} page={page} onChange={handleChangePage} />
      </Box>
    </div>
  );
};

export default connector(CategoryList);
