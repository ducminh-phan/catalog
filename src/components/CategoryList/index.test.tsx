import React from "react";

import { FetchCategoriesResponse } from "actions/category";
import { CATEGORIES_PER_PAGE } from "enums";
import Request from "utils/api";
import { fireEvent, render, setupUser, wait } from "utils/tests";

import CategoryList from "./index";

it("renders the list of categories correctly", async () => {
  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<FetchCategoriesResponse> =>
      Promise.resolve({
        totalCategories: 3,
        categories: [1, 2].map((id) => ({
          id,
          name: `name ${id}`,
          description: `description ${id}`,
        })),
      }),
  );

  // Render the first page
  const { baseElement, getByText, getAllByText } = render(<CategoryList />);
  await wait(() => getByText("name 1"));
  expect(baseElement).toMatchSnapshot();
  expect(getAllByText(/name.*/)).toHaveLength(2);
  expect(Request.get).toHaveBeenNthCalledWith(
    1,
    `categories?offset=0&limit=${CATEGORIES_PER_PAGE}`,
  );

  // We can navigate to the next page
  const nextPageButton = getByText("2");
  expect(nextPageButton).toBeInTheDocument();

  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<FetchCategoriesResponse> =>
      Promise.resolve({
        totalCategories: 3,
        categories: [
          {
            id: 3,
            name: `name 3`,
            description: `description 3`,
          },
        ],
      }),
  );

  await wait(() => fireEvent.click(nextPageButton));
  expect(baseElement).toMatchSnapshot();
  expect(getAllByText(/name.*/)).toHaveLength(1);
  expect(Request.get).toHaveBeenLastCalledWith(
    `categories?offset=${CATEGORIES_PER_PAGE}&limit=${CATEGORIES_PER_PAGE}`,
  );
});

it("show the button to add category to logged in users", async () => {
  setupUser();
  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<FetchCategoriesResponse> =>
      Promise.resolve({
        totalCategories: 3,
        categories: [1, 2].map((id) => ({
          id,
          name: `name ${id}`,
          description: `description ${id}`,
        })),
      }),
  );

  // Render the first page
  const { getByTestId, getByText } = render(<CategoryList />);
  await wait(() => getByText("name 1"));
  expect(getByTestId("add-category")).toBeInTheDocument();
});
