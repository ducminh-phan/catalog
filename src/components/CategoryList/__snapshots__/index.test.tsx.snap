// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the list of categories correctly 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-container"
          >
            <div
              class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            >
              <div
                class="MuiListItemText-root MuiListItemText-multiline"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  name 1
                </span>
                <p
                  class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  description 1
                </p>
              </div>
            </div>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-text"
                href="/categories/1"
                role="button"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 17l5-5-5-5v10z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </div>
          </li>
          <li
            class="MuiListItem-container"
          >
            <div
              class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            >
              <div
                class="MuiListItemText-root MuiListItemText-multiline"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  name 2
                </span>
                <p
                  class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  description 2
                </p>
              </div>
            </div>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-text"
                href="/categories/2"
                role="button"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 17l5-5-5-5v10z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </div>
          </li>
        </ul>
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-container"
          >
            <div
              class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            >
              <nav
                aria-label="pagination navigation"
                class="MuiPagination-root"
              >
                <ul
                  class="MuiPagination-ul"
                >
                  <li>
                    <button
                      aria-label="Go to previous page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-disabled MuiButtonBase-disabled"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                        />
                      </svg>
                    </button>
                  </li>
                  <li>
                    <button
                      aria-current="true"
                      aria-label="page 1"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-selected"
                      tabindex="0"
                      type="button"
                    >
                      1
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to page 2"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      2
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to next page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <div />
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders the list of categories correctly 2`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-container"
          >
            <div
              class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            >
              <div
                class="MuiListItemText-root MuiListItemText-multiline"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  name 3
                </span>
                <p
                  class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  description 3
                </p>
              </div>
            </div>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-text"
                href="/categories/3"
                role="button"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 17l5-5-5-5v10z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </div>
          </li>
        </ul>
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-container"
          >
            <div
              class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            >
              <nav
                aria-label="pagination navigation"
                class="MuiPagination-root"
              >
                <ul
                  class="MuiPagination-ul"
                >
                  <li>
                    <button
                      aria-label="Go to previous page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to page 1"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      1
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-current="true"
                      aria-label="page 2"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-selected"
                      tabindex="0"
                      type="button"
                    >
                      2
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to next page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-disabled MuiButtonBase-disabled"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                        />
                      </svg>
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <div />
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
